### DISCUSSION TITLE
Is it possible to repurpose the GIssueMarkerView for a more general decoration purpose? Instead of just for validation markers

### DISCUSSION THREAD

**Discussion by jmenzies12** [eclipse-glsp/glsp#1324] [Q&A] [answered]
Hey there,

I'm investigating the various features of GLSP and wanted to ask if it would be possible (I would be doing the work ofcourse) to repurpose the GIssueMarkerView used for validation icons for a more general decoration purpose? i.e. supply our own icons

Essentially our existing application allows us to display icons as decorators so we can show the user at a glance that a particular node in the diagram is special in someway (for example, it has a related node elsewhere in our models not connected to it via an edge).

Cheers!
Jonathan

**Comment by planger** [comment, marked as answer]
Hi Jonathan,

absolutely, you can definitely use the general issue marker functionality for your own purposes too, and not just for the use case of decorating validation results.

You can add `GIssueMarker` elements into your GModel directly from your `GModelFactory` and they'll be shown just as normal markers. To define the rendering (other icons, etc.) of your custom markers, you may need to bind your own issue marker view implementation. See [issue marker view implementation](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/views/issue-marker-view.tsx) and [binding your own views](https://eclipse.dev/glsp/documentation/rendering/) for more details.

If you don't want them to be a real part of the GModel (ie adding them to the GModel on the server directly), you can apply a similar strategy as GLSP core uses for transferring validation result markers down to the client and showing them. In the validation result marker implementation of GLSP, the markers are not added as part of the GModel, but added via a [feedback emitter](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/validation/validate.ts#L42) as a result of the `ApplyMarkersAction`. See [implementation](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/validation/validate.ts) for more details. You can of course introduce custom actions and replicate this approach with a similar custom mechanism.

Alternatively, you can also just reuse the existing mechanism above and just send an `ApplyMarkersAction` from your server at any time to show your own markers. The existing [validation feedback emitter](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/validation/validate.ts#L42) would react to this action and transfer them as feedback to the client GModel.

I hope this helps!

**Reply by jmenzies12** [reply]
Thank you for the really detailed and helpful answer!
